TESTS/mbedmicro-rtos-mbed/mail/main.cpp

```diff
@@ -15,69 +15,457 @@
-            result = result && expected_values;
-            const char *result_msg = expected_values ? "OK" : "FAIL";
-            printf("%3d %.2fV %.2fA ... [%s]\r\n", mail->counter,
-                                                   mail->voltage,
-                                                   mail->current,
-                                                   result_msg);
+            const uint8_t id = mail->thread_id;
+
+            // verify thread id
+            result = result && ((id == THREAD_1_ID) || (id == THREAD_2_ID) || (id == THREAD_3_ID));
+            // verify sent data
+            result = result && mail->data == data[id - 1]++;
             mail_box.free(mail);
+
             if (result == false || ++result_counter == QUEUE_SIZE) {
                 break;
             }
         }
     }
-    GREENTEA_TESTSUITE_RESULT(result);
-    return 0;
+    TEST_ASSERT_EQUAL(true, result);
+}
+
+/** Test message memory deallocation with block out of the scope
+
+    Given an empty mailbox
+    When try to free out of the scope memory block
+    Then it return appropriate error code
+ */
+void test_free_wrong()
+{
+    osStatus status;
+    Mail<uint32_t, 4> mail_box;
+    uint32_t *mail, data;
+
+    mail = &data;
+    status = mail_box.free(mail);
+    TEST_ASSERT_EQUAL(osErrorParameter, status);
+
+    mail = mail_box.alloc();
+    TEST_ASSERT_NOT_EQUAL(NULL, mail);
+
+    mail = &data;
+    status = mail_box.free(mail);
+    TEST_ASSERT_EQUAL(osErrorParameter, status);
+}
+
+/** Test message memory deallocation with null block
+
+    Given an empty mailbox
+    When try to free null ptr
+    Then it return appropriate error code
+ */
+void test_free_null()
+{
+    osStatus status;
+    Mail<uint32_t, 4> mail_box;
+    uint32_t *mail;
+
+    mail = NULL;
+    status = mail_box.free(mail);
+    TEST_ASSERT_EQUAL(osErrorParameter, status);
+
+    mail = mail_box.alloc();
+    TEST_ASSERT_NOT_EQUAL(NULL, mail);
+
+    mail = NULL;
+    status = mail_box.free(mail);
+    TEST_ASSERT_EQUAL(osErrorParameter, status);
+}
+
+/** Test same message memory deallocation twice
+
+    Given an empty mailbox
+    Then allocate message memory
+    When try to free it second time
+    Then it return appropriate error code
+ */
+void test_free_twice()
+{
+    osStatus status;
+    Mail<uint32_t, 4> mail_box;
+
+    uint32_t *mail = mail_box.alloc();
+    TEST_ASSERT_NOT_EQUAL(NULL, mail);
+
+    status = mail_box.free(mail);
+    TEST_ASSERT_EQUAL(osOK, status);
+
+    status = mail_box.free(mail);
+    TEST_ASSERT_EQUAL(osErrorResource, status);
+}
+
+/** Test get from empty mailbox with timeout set
+
+    Given an empty mailbox
+    When @a get is called on the mailbox with timeout of 50
+    Then mailbox returns status of osOK, but no data after specified amount of time
+ */
+void test_get_empty_timeout()
+{
+    Mail<uint32_t, 4> mail_box;
+
+    uint32_t start = us_ticker_read();
```

Comment:
Please use C++ API - there are tickers/timers etc, instead of using directly HAL